### PR TITLE
reload everything when a file changes

### LINKED ISSUE #835: Hot reloading does not reload all files
(Created from #823)

While hot reloading will reload a changed file, it does not appear that changes are actually used. For example, a changed file in the `paths/` directory will be reloaded. However, those changes have no effect until the `_.context.ts` file is reloaded as well.

### Setup
Have a file structure like so:
```
counterfact/
    paths/
        _mocks/
            my-custom-code.ts
        _.context.ts
```

In `_.context.ts`:
```ts
import { CustomClass } from './_mocks/my-custom-code';

console.log("_.context.ts loaded");
// Do stuff with CustomClass
```

In `./_mocks/my-custom-code`:

```ts
console.log("Custom code loaded");

const value = "Some value that's used in generating a mock response";

class CustomClass {
  // Use value in here somehow as part of generating a response
}

export { CustomClass }
```
### Trigger

1. Run counterfact (ex: `npx counterfact spec.openapi.yaml counterfact`)
2. Call counterfact api, see the `"Some value..."` in the HTTP response
3. Modify the custom code string to be something else (`"const value = "Changed!"`)
4. Note that `Custom code loaded` is logged
5. Make a request to counterfact. Note that the new value is not returned

However, if I trivially modify the `_.context.ts` file (touch it, add a comment, etc), I do see counterfact log the `_.context.ts loaded` message. If I then make a new HTTP request, I see the new value in the mock http response.

### PR DESCRIPTION
**Pull request by pmcelhaney**
Fixes #835 

When any file is touched, the entire app is reloaded. This is necessary because any file that imports the file that was changed (directly or indirectly) needs to be reloaded.

I'm a bit concerned this is going to cause performance problems on large codebases, but we'll see. Ideally, instead of reloading everything, we would trace dependencies and only reload the files that need to be reloaded.

### CODE REVIEW

**Comment by dethell** [comment]
> Fixes #835
> 
> When any file is touched, the entire app is reloaded. This is necessary because any file that imports the file that was changed (directly or indirectly) needs to be reloaded.
> 
> I'm a bit concerned this is going to cause performance problems on large codebases, but we'll see. Ideally, instead of reloading everything, we would trace dependencies and only reload the files that need to be reloaded.

I'm going to pull this branch and test against our codebase which is pretty large. Should know the results here in the next hour.

**Comment by dethell** [comment]
> I'm going to pull this branch and test against our codebase which is pretty large. Should know the results here in the next hour.

I can't boot up with this change. I'll PM you a sample openapi file to see if it fails on your end as well.

File: src/server/module-loader.ts [hunk 1]
```diff
@@ -167,50 +120,38 @@ export class ModuleLoader extends EventTarget {
...
-        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
-        this.registry.add(url, endpoint as Module);
-      }
-    } catch (error: unknown) {
-      if (
-        String(error) ===
-        "SyntaxError: Identifier 'Context' has already been declared"
-      ) {
-        // Not sure why Node throws this error. It doesn't seem to matter.
-        return;
```
Comment (pmcelhaney) [inline src/server/module-loader.ts:210]: @dethell I wonder if we need to put this back in. 

Actually looking at it with fresh eyes, I bet the imported file declares `Context` twice. And what will happen if there's a syntax error in one of the files? I think I may have accidentally removed some fault tolerance with recent refactoring.

**Comment by pmcelhaney** [comment]
> > I'm going to pull this branch and test against our codebase which is pretty large. Should know the results here in the next hour.
> 
> I can't boot up with this change. I'll PM you a sample openapi file to see if it fails on your end as well.

Did you check `main` to make sure it's broken in this change and not another recent change?

**Comment by dethell** [comment]
Here are some findings using `DEBUG=counterfact:*`:

It looks like it is stuck in an endless reloading loop. It will say "loading: /Users/davidethell/source/whatever/mock-server/api/.cache' over and over and then say "finished loading: " with the same path and just keep going back and forth between the two. Clearing the `.cache` folder doesn't help.

**Comment by dethell** [comment]
> Thanks. That log message wasn't very useful because it was logging the root directory instead of the subdirectory that's being loaded. Pushed an update. Try it now.

Trying now. interestingly enough, it did finally finish booting up and showing the REPL prompt. Not sure how long it took, maybe 20 minutes? I was on a meeting not watching it closely, but it did load.

Your update you pushed does show more detail and it confirms it just keeps looping through the same paths.

**Comment by pmcelhaney** [comment]
I just pushed an update that ignores files with syntax errors. I was hoping that would be the fix. 

I think the issue may be that watch() ends up calling load() again for every single file.

**Comment by dethell** [comment]
> I just pushed an update that ignores files with syntax errors. I was hoping that would be the fix.
> 
> I think the issue may be that watch() ends up calling load() again for every single file.

Still looping as before.

**Comment by pmcelhaney** [comment]
Closing this PR. I'm going to trace dependencies and reload only the affected files instead.

**Pull request closed without merging**